File: student-management/src/main/java/com/example/student_management/services/CourseService.java
package com.example.student_management.services;


import com.example.student_management.dto.CourseDto;

import com.example.student_management.dto.CreateCourseDto;

import com.example.student_management.entity.Course;

import com.example.student_management.entity.Semester;

import com.example.student_management.repository.CourseRepository;

import com.example.student_management.repository.SemesterRepository;

import org.modelmapper.ModelMapper;

import org.springframework.stereotype.Service;


import java.util.ArrayList;

import java.util.List;

import java.util.stream.Collectors;


@Service
public class CourseService {

    private final CourseRepository courseRepository;

    private final ModelMapper modelMapper;

    private final SemesterRepository semesterRepository;


    public CourseService(CourseRepository courseRepository, ModelMapper modelMapper, SemesterRepository semesterRepository) {
        this.courseRepository = courseRepository;

        this.modelMapper = modelMapper;

        this.semesterRepository = semesterRepository;

    }

    public CourseDto createCourse(CreateCourseDto createCourseDto) {
        Course course = modelMapper.map(createCourseDto, Course.class);

        Course savedCourse = courseRepository.save(course);


        List< Semester> semesterList = new ArrayList<>();


        for (int i=1;
 i<=createCourseDto.getTotalSemesters();
 i++){
            Semester semester = new Semester();

            semester.setSemester(i);

            semester.setCourse(savedCourse);

            semesterList.add(semesterRepository.save(semester));

        }


        savedCourse.setSemesters(semesterList);

        return modelMapper.map(savedCourse, CourseDto.class);

    }

    public List<CourseDto> getAllCourses() {
        return courseRepository.findAll().stream()
                .map(course -> modelMapper.map(course, CourseDto.class))
                .collect(Collectors.toList());

    }

    public CourseDto getCourseById(Long id) {
        Course course = courseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Course not found with id " + id));

        return modelMapper.map(course, CourseDto.class);

    }
}

File: nodesms/services/courseService.js
const { Course, Semester } = require('../models');

exports.createCourse = async (createCourseDto) => {
  const course = await Course.create(createCourseDto);

  const semesters = [];
  for (let i = 1; i <= createCourseDto.totalSemesters; i++) {
    const semester = await Semester.create({
      semester: i,
      courseId: course.id,
    });
    semesters.push(semester);
  }

  course.semesters = semesters;

  return course;
};

exports.getAllCourses = async () => {
  return await Course.findAll({ include: Semester });
};

exports.getCourseById = async (id) => {
  const course = await Course.findByPk(id, { include: Semester });
  if (!course) throw new Error('Course not found');
  return course;
};
